File: tcpdump/installation/installation_page.py
# Copyright 2021 BlueCat Networks (USA) Inc. and its affiliates. All Rights Reserved.
import os
import sys
import traceback
from flask import render_template, g, make_response, jsonify, request
from flask_restplus import Resource
from bluecat import route, util
import config.default_config as config
from main_app import app, api
from ...common.exceptions import UserException, InvalidParam
from ...common.tcpdump import TcpDump, InstallationResponse
from ...common.common import get_bam_name, validate_configuration_and_server
from .libs.models import installation_models


def module_path():
    encoding = sys.getfilesystemencoding()
    return os.path.dirname(os.path.abspath(__file__))


installation_tcpdump_ns = api.namespace(
    'installation',
    path='/installation',
    description='Installation TcpDump by Gateway',
)


@route(app, '/installation/installation_endpoint')
@util.workflow_permission_required('installation_page')
@util.exception_catcher
@util.ui_secure_endpoint
def installation_installation_page():
    return render_template(
        'installation_page.html',
        text=util.get_text(module_path(), config.language),
        options=g.user.get_options(),
        bam_name=get_bam_name(),
        bam_ip=g.user.get_api_netloc()
    )


@installation_tcpdump_ns.route('/package/uninstall')
class UninstallTcpDump(Resource):
    @util.workflow_permission_required('installation_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    @installation_tcpdump_ns.expect(installation_models.installation_tcpdump_model)
    def post(self):
        """ Uninstall the tcpdump"""
        try:
            message = []
            hosts = []
            data = request.get_json()
            configuration_name = data.get("configuration_name")
            server_ips = data.get("server_ips")
            for server_ip in server_ips:
                host = validate_configuration_and_server(configuration_name, server_ip)
                hosts.append(host)

            for host in hosts:
                try:
                    tcpdump = None
                    tcpdump = TcpDump(host=host)
                    response = tcpdump.uninstall_package()
                    tcpdump.disconnect()
                except Exception as e:
                    if tcpdump:
                        tcpdump.disconnect()
                    g.user.logger.error(traceback.format_exc())
                    response = InstallationResponse(host=host,
                                                    status=False,
                                                    message=str(e))
                message.append(response.__dict__)
            g.user.logger.debug(f"{request.url} {request.method}: Return state server-{message}")
            return jsonify(message), 200
        except InvalidParam as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 400)

        except Exception as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 500)


@installation_tcpdump_ns.route('/configurations')
class ConfigurationCollection(Resource):
    @util.rest_workflow_permission_required('installation_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self):
        """ Get all known configuration(s). """
        try:
            configurations = g.user.get_api().get_configurations()
            result = [config_entity.to_json() for config_entity in configurations]
            g.user.logger.debug(f"{request.url} {request.method}: Configuration list-{result}")
            return jsonify(result), 200
        except Exception as e:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)


@installation_tcpdump_ns.route('/configurations/<string:configuration>')
class Configuration(Resource):
    @util.rest_workflow_permission_required('installation_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self, configuration):
        """ Get configuration with specified name. """
        try:
            configuration = g.user.get_api().get_configuration(configuration)
            result = configuration.to_json()
            g.user.logger.debug(f"{request.url} {request.method}: Configuration-{result}")
            return jsonify(result), 200
        except Exception as e:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)


@installation_tcpdump_ns.route('/configurations/<string:configuration>/servers')
class ServersList(Resource):
    @util.workflow_permission_required('installation_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self, configuration):
        """ Get all known Server(s) with specified Configuration name. """
        try:
            configuration = g.user.get_api().get_configuration(configuration)
            servers = configuration.get_servers()
            servers = [server.to_json() for server in servers]
            g.user.logger.debug(f"{request.url} {request.method}: Server list-{servers}")
            return jsonify(servers), 200
        except UserException as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 400)
        except Exception as ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(ex)), 500)

File: tcpdump/installation/js/installation_page.js
// Copyright 2021 BlueCat Networks (USA) Inc. and its affiliates. All Rights Reserved.
var dataHTML = []

$(document).ready(function () {
    getConfigurations();
    loadServer();
    resetDataTable();
});


function getConfigurations() {
    $.ajax({
        url: `/api/v1/installation/configurations`,
        type: 'get',
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            var result = [];
            result.push(`<option value='Please Select'>Please Select</option>`)
            for (i = 0; i < data.length; i++) {
                result.push(`<option value='${data[i]['name']}'>${data[i]['name']}</option>`)
            }
            $('#configuration').html(result);
            getServers();
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}


function getServers() {
    var configName = $('#configuration').val();
    $('#server').html(`<option value='${BAM_NAME}(${BAM_IP})'>${BAM_NAME}(${BAM_IP})</option>`)
    if (configName.includes('Please Select')) {
        $('#server').val(BAM_NAME + '(' + BAM_IP + ')');
//        $('#install').removeAttr('disabled', 'disabled');
//        $('#uninstall').removeAttr('disabled', 'disabled');
        $('#remove').removeAttr('disabled', 'disabled');
        return false
    } else {
        loadServer();
//        $('#install').attr('disabled', true);
//        $('#uninstall').attr('disabled', true);
        $('#remove').attr('disabled', true);
    }

    $.ajax({
        url: `/api/v1/installation/configurations/${configName}/servers`,
        type: 'get',
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            var result = [];
            for (i = 0; i < data.length; i++) {
                serverIp = getProperty(data[i], 'defaultInterfaceAddress')
                result.push(`<option value='${data[i]['name']}(${serverIp})'>${data[i]['name']}(${serverIp})</option>`)
            }
            $('#server').html($('#server').html() + result);
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}
//
//function installTCPDump() {
//    var configName = $('#configuration').val();
//    if (configName.includes('Please Select')){
//        configName = '';
//    }
//    var element = '';
//    var servers = [];
//    var listServerIp = [];
//    servers = $('#server').val();
//    for (var i = 0; i < servers.length; i++) {
//        serverIp = servers[i].substring(servers[i].indexOf('(') + 1,servers[i].indexOf(')'))
//        listServerIp.push(serverIp)
//    }
//    var data = {
//        "configuration_name": configName,
//        "server_ips": listServerIp
//    }
//    $.ajax({
//        url: `/api/v1/installation/package/install`,
//        type: 'post',
//        data: JSON.stringify(data),
//        dataType: 'json',
//        contentType: 'application/json',
//        success: function (data) {
//            dataHTML = []
//            for (i = 0; i < data.length; i++) {
//                if (data[i].status == true) {
//                    dataHTML.push(`<tr class="rowTable"><td>${data[i].server_name}</td>
//                                 <td style="color:#76CE66;">${data[i].message}</td></tr>`)
//                } else {
//                    dataHTML.push(`<tr class="rowTable"><td>${data[i].server_name}</td>
//                                 <td style="color:#FF0000;">${data[i].message}</td></tr>`)
//                }
//                $('#dataTable > #bodyTable').html(dataHTML);
//            }
//        },
//        error: function (message) {
//            definedErrorMessage(message)
//        }
//    });
//}
//
//function uninstallTCPDump() {
//    var configName = $('#configuration').val();
//    if (configName.includes('Please Select')){
//        configName = '';
//    }
//    var element = '';
//    var servers = [];
//    var listServerIp = [];
//    servers = $('#server').val();
//    for (var i = 0; i < servers.length; i++) {
//        serverIp = servers[i].substring(servers[i].indexOf('(') + 1,servers[i].indexOf(')'))
//        listServerIp.push(serverIp)
//    }
//    var data = {
//        "configuration_name": configName,
//        "server_ips": listServerIp
//    }
//    $.ajax({
//        url: `/api/v1/installation/package/uninstall`,
//        type: 'post',
//        data: JSON.stringify(data),
//        dataType: 'json',
//        contentType: 'application/json',
//        success: function (data) {
//            for (i = 0; i < data.length; i++) {
//                if (data[i].status == true) {
//                    element = '<div class="row" data-region-id="' + 'regionName' + '" style="margin-bottom: 10px;">' +
//                               '<div class="col-xs-4" style="padding-left: 14px;width:30%;overflow-wrap: break-word;">' +
//                               data[i].server_name+
//                              '</div>' +
//                              '<div class="col-xs-6" style="color:#76CE66;margin-left:-21px;overflow-wrap: break-word">' +
//                               data[i].message+
//                              '</div>'+
//                              '</div>'
//                } else {
//                    element = '<div class="row" data-region-id="' + 'regionName' + '" style="margin-bottom: 10px;">' +
//                               '<div class="col-xs-4" style="padding-left: 14px;width:30%;overflow-wrap: break-word;">' +
//                               data[i].server_name+
//                              '</div>' +
//                              '<div class="col-xs-6" style="color:#FF0000;margin-left:-21px;overflow-wrap: break-word">' +
//                               data[i].message+
//                              '</div>'+
//                              '</div>'
//                }
//                $('#tcpdump-table-body').append(element)
//            }
//        },
//        error: function (message) {
//            definedErrorMessage(message)
//        }
//    });
//}

function removeTCPDump() {
    var configName = $('#configuration').val();
    if (configName.includes('Please Select')){
        configName = '';
    }
    var element = '';
    var servers = [];
    var listServerIp = [];
    servers = $('#server').val();
    for (var i = 0; i < servers.length; i++) {
        serverIp = servers[i].substring(servers[i].indexOf('(') + 1,servers[i].indexOf(')'))
        listServerIp.push(serverIp)
    }
    var data = {
        "configuration_name": configName,
        "server_ips": listServerIp
    }
    $.ajax({
        url: `/api/v1/installation/package/uninstall`,
        type: 'post',
        data: JSON.stringify(data),
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            dataHTML = []
            for (i = 0; i < data.length; i++) {
                if (data[i].status == true) {
                    dataHTML.push(`<tr class="rowTable"><td>${data[i].server_name}</td>
                                 <td style="color:#76CE66;">${data[i].message}</td></tr>`)
                } else {
                    dataHTML.push(`<tr class="rowTable"><td>${data[i].server_name}</td>
                                 <td style="color:#FF0000;">${data[i].message}</td></tr>`)
                }
                $('#dataTable > #bodyTable').html(dataHTML);
            }
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}

function getProperty(entity, name) {
    var result = "";
    var properties = entity['properties'].split("|");
    for (p = 0; p < properties.length; p++) {
        if (properties[p].includes(`${name}`)) {
            result = properties[p].split("=")[1];
        }
    }
    return result
}

function loadServer() {
    $('#server').select2({
        placeholder: 'Please Select',
        minimumResultsForSearch: -1,
        allowClear: true,
        templateResult: hideSelected,
        createTag: function () {
            return null;
        }
    });
}

function hideSelected(value) {
    if (value && !value.selected) {
        return $('<span>' + value.text + '</span>');
    }
}

function disabledButton(){
    servers = $('#server').val();
    if (servers.length > 0) {
//        $('#install').removeAttr('disabled', 'disabled');
//        $('#uninstall').removeAttr('disabled', 'disabled');
        $('#remove').removeAttr('disabled', 'disabled');
    } else {
//        $('#install').attr('disabled', true);
//        $('#uninstall').attr('disabled', true);
        $('#remove').attr('disabled', true);
        resetDataTable();
    }
}

function resetDataTable(){
    dataHTML = []
    dataHTML.push(`<tr class="noData"><td colspan="2" style="font-weight:bold">There is no data in this table</td></tr>`);
    $('#dataTable > #bodyTable').html(dataHTML);
}